rust/main/chains/hyperlane-starknet/src/ism/multisig_ism.rs

```diff
@@ -0,0 +1,148 @@
+
+use crate::contracts::multisig_ism::{
+    Bytes as StarknetBytes, Message as StarknetMessage, MultisigIsm as StarknetMultisigIsmInternal,
+};
+use crate::error::HyperlaneStarknetError;
+use crate::types::HyH256;
+use crate::{build_single_owner_account, ConnectionConf, Signer, StarknetProvider};
+
+impl<A> std::fmt::Display for StarknetMultisigIsmInternal<A>
+where
+    A: starknet::accounts::ConnectedAccount + Sync + std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+/// A reference to a MultisigISM contract on some Starknet chain
+#[derive(Debug)]
+#[allow(unused)]
+pub struct StarknetMultisigIsm {
+    contract: Arc<StarknetMultisigIsmInternal<SingleOwnerAccount<AnyProvider, LocalWallet>>>,
+    provider: StarknetProvider,
+    conn: ConnectionConf,
+}
+
+impl StarknetMultisigIsm {
+    /// Create a reference to a MultisigISM at a specific Starknet address on some
+    /// chain
+    pub fn new(
+        conn: &ConnectionConf,
+        locator: &ContractLocator,
+        signer: Signer,
+    ) -> ChainResult<Self> {
+        let account = build_single_owner_account(
+            &conn.url,
+            signer.local_wallet(),
+            &signer.address,
+            false,
+            locator.domain.id(),
+        );
+
+        let ism_address: FieldElement = HyH256(locator.address)
+            .try_into()
+            .map_err(HyperlaneStarknetError::BytesConversionError)?;
+
+        let contract = StarknetMultisigIsmInternal::new(ism_address, account);
+
+        Ok(Self {
+            contract: Arc::new(contract),
+            provider: StarknetProvider::new(locator.domain.clone(), conn),
+            conn: conn.clone(),
+        })
+    }
+
+    #[allow(unused)]
+    pub fn contract(
+        &self,
+    ) -> &StarknetMultisigIsmInternal<SingleOwnerAccount<AnyProvider, LocalWallet>> {
+        &self.contract
+    }
+}
+
+impl HyperlaneChain for StarknetMultisigIsm {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.provider.domain()
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl HyperlaneContract for StarknetMultisigIsm {
+    fn address(&self) -> H256 {
+        HyH256::from(self.contract.address).0
+    }
+}
+
+impl From<&HyperlaneMessage> for StarknetMessage {
+    fn from(message: &HyperlaneMessage) -> Self {
+        StarknetMessage {
+            version: message.version,
+            nonce: message.nonce,
+            origin: message.origin,
+            sender: StarknetU256::from_bytes_be(&message.sender.to_fixed_bytes()),
+            destination: message.destination,
+            recipient: StarknetU256::from_bytes_be(&message.recipient.to_fixed_bytes()),
+            body: StarknetBytes {
+                size: message.body.len() as u32,
+                data: message.body.iter().map(|b| *b as u128).collect(),
+            },
+        }
+    }
+}
```

Comment:
It looks like we have this kind of conversion implemented is many places. I wonder if it would be possible to avoid code duplication?

Reply:
yes well that's a whole point of having it under a From implementation

Reply:
as I mentionned in a previous comment, unfortunately due to the way cainome (abigen for starknet) works we have to re-write this everytime because the underlying struct is not the same